import os
import glob
import pandas as pd
from scripts.resources import Resources
include: "scripts/general_functions.smk"
from snakemake.utils import min_version

# set minimum snakemake version
min_version("7.25.0")

# load config file
configfile: "config/config.yaml"

# Load sample information
csv = pd.read_csv("config/samples.csv")

# load genome resources to be used in rules
resources = Resources(config["genome"], config["ensembl_genome_build"])

# get sample directories
DIRS = dirs()
SAMPLES = samples()

# import rules
include: "rules/fastqc.smk"
include: "rules/trimming.smk"
include: "rules/resources.smk"
include: "rules/setup.smk"
include: "rules/damid.smk"
include: "rules/bedgraph_processing.smk"
#include: "rules/plotting.smk"

# target rule
rule all:
    input: 
        targets()


# save snakemake terminal output to log file
snake_log = "logs/snakemake/snakemake.log"
os.makedirs("logs/snakemake", exist_ok=True)

onsuccess: 
    shell("cp -v {log} {snake_log}")
    shell(f"pigz {resources.fasta}") # compress genome files
    print("Analysis finished successfully!")

onerror:
    shell("cp -v {log} {snake_log}")
    print(f"Analysis (partly) failed...\nCheck {snake_log} for details")


